### PR TITLE
Slicing API updates

### PR DESCRIPTION
**Pull request by vincentschen**
**Changes**
- [x] Add `PandasSlicer` for visualization
```
pd_slicer = PandasSlicer(df_valid)
polarity_df = pd_slicer.slice(textblob_polarity)
polarity_df[["text", "label"]].head()
```

- [x] Add `SliceScorer`
```
scorer = Scorer(metrics=["accuracy", "f1"])
scorer = SliceScorer(slice_names, scorer)
scorer.score(
    S_matrix=S_test,
    golds=Y_test,
    preds=preds_test,
    probs=preds_to_probs(preds_test, 2),
    as_dataframe=True,
)
```
- [x] Add dataframe visualizers to existing methods
```
slice_model.score([dl_valid], as_dataframe=True)
```

**Bug fixes**
- ~~Fix `slicing_function` alias to `labeling_function` decorator (currently initializes a `LabelingFunction`)~~ Moving to a separate PR

**Test Plan**
* Add associated tests for above changes
* `tox` + `tox -e complex` passes

### CODE REVIEW

File: snorkel/classification/utils.py [hunk 1]
```diff
@@ -116,3 +117,17 @@ def collect_flow_outputs_by_suffix(
         for flow_name in sorted(flow_dict.keys())
         if flow_name.endswith(suffix)
     ]
+
+
+def metrics_dict_to_dataframe(metrics_dict: Dict[str, float]) -> pd.DataFrame:
```
Comment (paroma) [inline snorkel/classification/utils.py]: should this have a task field as well? (or I may be remembering an older format of the metric dictionary)

File: snorkel/slicing/monitor.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+SliceMetricsDict = Dict[str, Dict[str, float]]
+
+
+class PandasSlicer:
+    """Create DataFrames corresponding to slices.
+
+    Parameters
+    ----------
+    df
+        A pandas DataFrame that will be sliced
```
Comment (paroma) [inline snorkel/slicing/monitor.py]: nit: add more description of what it'll contain (fields). high level okay

File: snorkel/slicing/monitor.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        S_matrix
+            An [num_examples x num_slices] matrix of slicing function outputs
+        golds
+            Gold (aka ground truth) labels (integers)
+        preds
+            Predictions (integers)
+        probs:
+            Probabilities (floats)
+        as_dataframe
+            A boolean indicating whether to return results as pandas DataFrame (True)
```
Comment (paroma) [inline snorkel/slicing/monitor.py]: nit: say what default is

File: snorkel/slicing/utils.py [hunk 4]
```diff
@@ -25,13 +25,13 @@ def add_slice_labels(
         A DictDataLoader whose dataset.Y_dict attribute will be modified in place
     base_task
        The Task for which we want corresponding slice tasks/labels
-    slice_labels
+       S_matrix
```
Comment (paroma) [inline snorkel/slicing/utils.py]: do we say `L_matrix` and `T_matrix`? or just `L` and `T`?

File: test/slicing/test_monitor.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        scorer = SliceScorer(slice_names, scorer)
+        metrics = scorer.score(S_matrix=S_matrix, golds=golds, preds=preds, probs=probs)
+        self.assertEqual(metrics["overall"]["accuracy"], 0.6)
+        self.assertEqual(metrics["sf"]["accuracy"], 2.0 / 3.0)
+
+        metrics_df = scorer.score(
+            S_matrix=S_matrix, golds=golds, preds=preds, probs=probs, as_dataframe=True
+        )
+        self.assertTrue(isinstance(metrics_df, pd.DataFrame))
+        self.assertEqual(metrics_df["accuracy"]["overall"], 0.6)
```
Comment (paroma) [inline test/slicing/test_monitor.py]: super nit: if there are other fields, can use to check those here

Comment (vincentschen) [reply test/slicing/test_monitor.py]: The full dict should have accuracy -> [overall, sf] as keys!

File: snorkel/classification/snorkel_classifier.py [hunk 6]
```diff
@@ -375,7 +375,10 @@ def predict(
 
     @torch.no_grad()
     def score(
-        self, dataloaders: List[DictDataLoader], remap_labels: Dict[str, str] = {}
+        self,
+        dataloaders: List[DictDataLoader],
+        remap_labels: Dict[str, str] = {},
+        as_dataframe: bool = False,
```
Comment (bhancock8) [inline snorkel/classification/snorkel_classifier.py]: Update docstring

File: snorkel/slicing/monitor.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+        S_matrix = PandasSFApplier([slicing_function]).apply(self.df)
+
+        # Index into the SF labels for the first (and only) column
+        df_idx = np.where(S_matrix[:, 0])[0]
+        return self.df.iloc[df_idx]
+
+
+class SliceScorer:
+    """Scorer that returns metrics on overall performance and slices.
```
Comment (bhancock8) [inline snorkel/slicing/monitor.py]: "...as well as slices."

Comment (bhancock8) [reply snorkel/slicing/monitor.py]: Just to emphasize that that's the difference here from a normal scorer.

File: snorkel/classification/utils.py (hunk 1 above)
Comment (vincentschen) [reply snorkel/classification/utils.py]: good catch, in this case, `task_name == label_name`

File: snorkel/slicing/utils.py (hunk 4 above)
Comment (vincentschen) [reply snorkel/slicing/utils.py]: I don't think there is a transformation matrix; but I'll change it to `S` for consistency with `L`

File: snorkel/slicing/monitor.py (hunk 7 above)
Comment (vincentschen) [reply snorkel/slicing/monitor.py]: 👍
